### PR TITLE
[WIP] pkp/pkp-lib#8885 Retrieve review assignments associated with submissions list

### PR DESCRIPTION
**Pull request by Vitaliy-1**

### CODE REVIEW

File: classes/submission/maps/Schema.php [hunk 1]
```diff
@@ -494,6 +612,26 @@ public function getPropertyStages(Submission $submission): array
         return $stages;
     }
 
+    /**
+     * Check if deciding editors are assigned to the submission
+     *
+     * @param Enumerable<StageAssignment> $stageAssignments
+     */
+    protected function getPropertyStageAssignments(Enumerable $stageAssignments): bool
```
Comment (defstat) [inline classes/submission/maps/Schema.php:620]: Can this be simplified to something like this 

```
protected function getPropertyStageAssignments(Enumerable $stageAssignments): bool
{
    return $stageAssignments->isNotEmpty() && $stageAssignments->contains(function ($stageAssignment) {
        return !$stageAssignment->recommendOnly;
    });
}
```

File: classes/submission/maps/Schema.php [hunk 2]
```diff
@@ -504,4 +642,54 @@ protected function getUserGroup(int $userGroupId): ?UserGroup
...
+        $submissionIds = $submissions->map(fn (Submission $submission) => $submission->getId())->toArray();
+        return StageAssignment::withSubmissionIds($submissionIds)
+            ->withRoleIds($roleIds)
+            ->get();
+    }
+
+    /**
+     * @return Collection<ReviewRound> [round => ReviewRound] sorted by the round number
+     */
+    protected function getReviewRoundsFromSubmission(Submission $submission): Collection
```
Comment (defstat) [inline classes/submission/maps/Schema.php:663]: Can this be simplified by something like this 

```
return collect($reviewRoundDao->getBySubmissionId($submission->getId())->toIterator())
            ->keyBy(fn ($reviewRound) => $reviewRound->getData('round'));
```

File: classes/submission/maps/Schema.php [hunk 3]
```diff
@@ -504,4 +642,54 @@ protected function getUserGroup(int $userGroupId): ?UserGroup
...
+        }
+
+        return $reviewRounds;
+    }
+
+    /**
+     * @return ?bool if there is one or more recommending editors assigned and all recommendations are given return true,
+     * otherwise returns false (recommendations are not yet given) or null (no recommending editors assigned)
+     */
+    protected function areRecommendationsIn(ReviewRound $currentReviewRound, Enumerable $stageAssignments): ?bool
```
Comment (defstat) [inline classes/submission/maps/Schema.php:678]: I suggest we use the one line conditional operator (? :) here

File: classes/submission/maps/Schema.php [hunk 4]
```diff
@@ -144,21 +195,50 @@ public function summarizeMany(Enumerable $collection, LazyCollection $userGroups
         $this->collection = $collection;
         $this->userGroups = $userGroups;
         $this->genres = $genres;
-        return $collection->map(function ($item) {
-            return $this->summarize($item, $this->userGroups, $this->genres);
-        });
+        $this->reviewAssignments = Repo::reviewAssignment()->getCollector()->filterBySubmissionIds($collection->keys()->toArray())->getMany()->remember();
```
Comment (defstat) [inline classes/submission/maps/Schema.php:198]: Here you are using the `remember()` function while in other places in the same file you are not (like `Repo::reviewAssignment()->getCollector()->filterBySubmissionIds([$item->getId()])->getMany()`). Is this indented?

File: classes/submission/maps/Schema.php [hunk 5]
```diff
@@ -105,11 +132,21 @@ public function map(Submission $item, LazyCollection $userGroups, array $genres)
...
+    public function summarize(
+        Submission $item,
+        LazyCollection $userGroups,
+        array $genres,
+        ?Enumerable $reviewAssignments = null,
+        ?Enumerable $stageAssignments = null
+    ): array {
         $this->userGroups = $userGroups;
         $this->genres = $genres;
+        $this->reviewAssignments = $reviewAssignments ?? Repo::reviewAssignment()->getCollector()->filterBySubmissionIds([$item->getId()])->getMany();
```
Comment (defstat) [inline classes/submission/maps/Schema.php:147]: Here the `$this->reviewAssignments` is a `LazyCollection`, while `$this->stageAssignments` right bellow is a `Collection`. Do you think that we could take any advantage into changing the second one to `LazyCollection` by changing the `getStageAssignmentsBySubmissions`'s get function to `lazy()` or something similar?

File: classes/submission/maps/Schema.php [hunk 6]
```diff
@@ -504,4 +642,54 @@ protected function getUserGroup(int $userGroupId): ?UserGroup
...
+     * @param Enumerable<Submission> $submissions
+     *
+     * @return Collection<StageAssignment> The collection of stage assignments associated with submissions
+     */
+    protected function getStageAssignmentsBySubmissions(Enumerable $submissions, array $roleIds = []): Collection
+    {
+        $submissionIds = $submissions->map(fn (Submission $submission) => $submission->getId())->toArray();
+        return StageAssignment::withSubmissionIds($submissionIds)
+            ->withRoleIds($roleIds)
+            ->get();
```
Comment (defstat) [inline classes/submission/maps/Schema.php:657]: See a comment above about the use of `lazy()` function here

File: classes/submission/maps/Schema.php (hunk 4 above)
Comment (Vitaliy-1) [reply classes/submission/maps/Schema.php:198]: Thanks, missed that!

File: classes/submission/maps/Schema.php (hunk 5 above)
Comment (Vitaliy-1) [reply classes/submission/maps/Schema.php:147]: In theory - yes, it should be more memory efficient.

**Comment by Vitaliy-1** [comment]
Thanks, @defstat! This is intended mainly for the new submissions list.

**Pull request merged**